codex/utils/safeasynciter.nim

```diff
@@ -0,0 +1,233 @@
+## - items - to iterate over the async iterator
+## - pairs - to iterate over the async iterator and return the index of each item
+## - mapAsync - to convert a regular sync iterator (Iter) to an async iter (SafeAsyncIter)
+## - map - to convert one async iterator (SafeAsyncIter) to another async iter (SafeAsyncIter)
+## - mapFilter - to convert one async iterator (SafeAsyncIter) to another async iter (SafeAsyncIter) and apply filtering at the same time
+## - filter - to filter an async iterator (SafeAsyncIter) returning another async iterator (SafeAsyncIter)
+## - delayBy - to delay each item returned by async iter by a given duration
+## - empty - to create an empty async iterator (SafeAsyncIter)
+
+type
+  SafeFunction[T, U] =
+    proc(fut: T): Future[U] {.async: (raises: [CancelledError]), gcsafe, closure.}
+  SafeIsFinished = proc(): bool {.raises: [], gcsafe, closure.}
+  SafeGenNext[T] = proc(): Future[T] {.async: (raises: [CancelledError]), gcsafe.}
+
+  SafeAsyncIter*[T] = ref object
+    finished: bool
+    next*: SafeGenNext[?!T]
+
+proc flatMap[T, U](
+    fut: auto, fn: SafeFunction[?!T, ?!U]
+): Future[?!U] {.async: (raises: [CancelledError]).} =
+  let t = await fut
+  await fn(t)
+
+proc flatMap[T, U](
+    fut: auto, fn: SafeFunction[?!T, Option[?!U]]
+): Future[Option[?!U]] {.async: (raises: [CancelledError]).} =
+  let t = await fut
+  await fn(t)
+
+########################################################################
+## SafeAsyncIter public interface methods
+########################################################################
+
+proc new*[T](
+    _: type SafeAsyncIter[T],
+    genNext: SafeGenNext[?!T],
+    isFinished: IsFinished,
+    finishOnErr: bool = true,
+): SafeAsyncIter[T] =
+  ## Creates a new Iter using elements returned by supplier function `genNext`.
+  ## Iter is finished whenever `isFinished` returns true.
+  ##
+
+  var iter = SafeAsyncIter[T]()
+
+  proc next(): Future[?!T] {.async: (raises: [CancelledError]).} =
+    try:
+      if not iter.finished:
+        let item = await genNext()
+        if finishOnErr and err =? item.errorOption:
+          iter.finished = true
+          return failure(err)
+        if isFinished():
+          iter.finished = true
+        return item
+      else:
+        return failure("SafeAsyncIter is finished but next item was requested")
+    except CancelledError as err:
+      iter.finished = true
+      raise err
+
+  if isFinished():
+    iter.finished = true
+
+  iter.next = next
+  return iter
+
+# forward declaration
+proc mapAsync*[T, U](
+  iter: Iter[T], fn: SafeFunction[T, ?!U], finishOnErr: bool = true
+): SafeAsyncIter[U]
+
+proc new*[U, V: Ordinal](
+    _: type SafeAsyncIter[U], slice: HSlice[U, V], finishOnErr: bool = true
+): SafeAsyncIter[U] =
+  ## Creates new Iter from a slice
+  ##
+
+  let iter = Iter[U].new(slice)
+  mapAsync[U, U](
+    iter,
+    proc(i: U): Future[?!U] {.async: (raises: [CancelledError]).} =
+      success[U](i),
+  )
```

Comment:
Shouldn't `mapAsync` have parameter `finishOnErr = finishOnErr`, like in the proc below?

Reply:
Yes, it should. Good catch! Fixed.